### PR TITLE
Add Version Checks on Para Upgrade

### PR DESCRIPTION
**Pull request by joepetrowski**
Adds checks to ensure that the spec version is increasing and that the spec name is the same. This will prevent accidentally upgrading a parachain to a different runtime.

### Migration

Although this changes storage, I did not increase the storage version or provide a migration. The reason is that the upgrade itself will `kill` the affected storage item, and the next time an upgrade is authorized, the new type will be written. But this will break any front ends that show authorized upgrades, so let me know if I should bump the storage version for this.

### Transaction Version

Unfortunately breaks it, although it's a Root call. Could also add a new function `authorize_upgrade_without_checks` as an alternative.

### CODE REVIEW

File: pallets/parachain-system/src/lib.rs [hunk 1]
```diff
@@ -134,6 +136,36 @@ impl CheckAssociatedRelayNumber for AnyRelayNumber {
 	fn check_associated_relay_number(_: RelayChainBlockNumber, _: RelayChainBlockNumber) {}
 }
 
+/// Information needed when a new runtime binary is submitted and needs to be authorized before
+/// replacing the current runtime.
+#[derive(Decode, Encode, Default, PartialEq, Eq, MaxEncodedLen)]
+struct CodeUpgradeAuthorization<T>
```
Comment (bkchr) [inline pallets/parachain-system/src/lib.rs:142]:
```suggestion
#[derive(Decode, Encode, Default, PartialEq, Eq, MaxEncodedLen, TypeInfo)]
#[scale_info(skip_type_params(T))]
struct CodeUpgradeAuthorization<T>
```

File: pallets/parachain-system/src/lib.rs [hunk 2]
```diff
@@ -134,6 +136,36 @@ impl CheckAssociatedRelayNumber for AnyRelayNumber {
...
+	fn type_info() -> scale_info::Type {
+		scale_info::Type::builder()
+			.path(scale_info::Path::new("CodeUpgradeAuthorization", module_path!()))
+			.composite(scale_info::build::Fields::unnamed().field(|f| {
+				f.ty::<u8>().docs(&[
+					"Raw code upgrade authorization bytes, encodes code_hash + check_version",
+				])
+			}))
+	}
+}
```
Comment (bkchr) [inline pallets/parachain-system/src/lib.rs:167]:
```suggestion
```

File: pallets/parachain-system/src/lib.rs [hunk 3]
```diff
@@ -442,17 +474,40 @@ pub mod pallet {
...
 
+		/// Provide the preimage (runtime binary) `code` for an upgrade that has been authorized.
+		///
+		/// If the authorization required a version check, this call will ensure the spec name
+		/// remains unchanged and that the spec version has increased.
+		///
+		/// Note that this function will not apply the new `code`, but only attempt to schedule the
+		/// upgrade with the Relay Chain.
+		///
+		/// The origin for this call should be unsigned.
```
Comment (bkchr) [inline pallets/parachain-system/src/lib.rs:510]:
```suggestion
		/// All origins are allowed.
```

We don't check the origin at all (on purpose).

File: pallets/parachain-system/src/lib.rs [hunk 4]
```diff
@@ -712,9 +774,28 @@ pub mod pallet {
...
+				.and_then(|v| RuntimeVersion::decode(&mut &v[..]).ok())
+				.ok_or(Error::<T>::FailedToExtractRuntimeVersion)?;
+
+			if new_version.spec_name != current_version.spec_name {
+				return Err(Error::<T>::InvalidSpecName.into())
+			}
+
+			if new_version.spec_version <= current_version.spec_version {
+				return Err(Error::<T>::SpecVersionNeedsToIncrease.into())
+			}
```
Comment (bkchr) [inline pallets/parachain-system/src/lib.rs:796]:
```suggestion
			frame_system::Pallet::<T>::can_set_code(code)?;
```

**Review by bkchr — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2468624

**Review by gilescope — APPROVED** [review]

**Review by ruseinov — APPROVED** [review]

**Comment by joepetrowski** [comment]
bot merge

**Comment by xlc** [comment]
Can we have some scary labels to ensure this PR is getting attentions by parachain teams? It is a breaking change and will impact scripts related to runtime upgrade.
It will be even better if you can just make it a non-breaking change.

**Pull request merged**